include/sacn/dmx_merger.h

```diff
@@ -0,0 +1,166 @@
+      that slot, or DMX_MERGER_SOURCE_INVALID to indicate that no source is providing values for that slot.
+      Memory is owned by the application.*/
+  source_id_t* slot_owners;
+
+} SacnDmxMergerConfig;
+
+/*!
+ * \brief An initializer for an SacnDmxMergerConfig struct.
+ *
+ * Usage:
+ * \code
+ * // Create the struct
+ * SacnDmxMergerConfig merger_config = SACN_DMX_MERGER_CONFIG_INIT;
+ * // Now fill in the members of the struct
+ * \endcode
+ *
+ */
+#define SACN_DMX_MERGER_CONFIG_INIT \
+  {                                 \
+    0, NULL, NULL                   \
+  }
+
+/*! The current input data for a single source of the merge.
+    This is exposed only for informational purposes, as the
+    application call a variant of sacn_dmx_merger_update_source */
+typedef struct SacnDmxMergerSource
+{
+  /*! The UUID (e.g. sACN CID) of the DMX source. */
+  EtcPalUuid cid;
+
+  /*! The DMX data values (0 - 255). */
+  uint8_t values[DMX_ADDRESS_COUNT];
+
+  /*! Some sources don't send all 512 values, so here's how much of values to use.*/
+  size_t valid_value_count;
+
+  /*! The sACN per-universe priority (0 - 255). */
+  uint8_t universe_priority;
+
+  /*! Whether or not the address_priority buffer is valid. */
+  bool address_priority_valid;
+
+  /*! The sACN per-address (startcode 0xdd) priority (1-255, 0 means not sourced).
+      If the source does not */
+  uint8_t address_priority[DMX_ADDRESS_COUNT];
+
+} SacnDmxMergerSource;
+
+etcpal_error_t sacn_dmx_merger_init();
+void sacn_dmx_merger_deinit(void);
+
+etcpal_error_t sacn_dmx_merger_create(const SacnDmxMergerConfig* config, sacn_dmx_merger_t* handle);
+etcpal_error_t sacn_dmx_merger_destroy(sacn_dmx_merger_t handle);
+
+etcpal_error_t sacn_dmx_merger_add_source(sacn_dmx_merger_t merger, const EtcPalUuid* source_cid,
+                                          source_id_t* source_id);
+etcpal_error_t sacn_dmx_merger_remove_source(sacn_dmx_merger_t merger, source_id_t source);
+const SacnDmxMergerSource* sacn_dmx_merger_get_source(sacn_dmx_merger_t merger, source_id_t source);
+etcpal_error_t sacn_dmx_merger_update_source_data(sacn_dmx_merger_t merger, source_id_t source,
+                                                  const uint8_t* new_values, size_t new_values_count, uint8_t priority,
+                                                  const uint8_t* address_priorities, size_t address_priorities_count);
+// TODO: If Receiver API changes to notify both values and per-address priority data in the same callback, this should
+// change!!
+etcpal_error_t sacn_dmx_merger_update_source_from_sacn(sacn_dmx_merger_t merger, source_id_t source,
+                                                       const SacnHeaderData* header, const uint8_t* pdata);
+etcpal_error_t sacn_dmx_merger_stop_source_per_address_priority(sacn_dmx_merger_t merger, source_id_t source);
+
+// TODO: Do we need this?
+etcpal_error_t sacn_dmx_merger_recalculate(sacn_dmx_merger_t merger);
```

Comment:
And not necessarily even then.  The update functions could be optimized to only need to calculate changes on a particular source, and only look at the other sources when that source has changed priority or given up control of the channels.  I seem to recall Dan's merger doing it this way.

This function would be a complete recalc.

Reply:
It sounds like we could go either way with this. We could either keep it in case anyone finds a use for it, or get rid of it and see if anyone brings up a need for it in a larger review.

Reply:
Let's keep it in until you've finished your rough implementation.  If the other functions don't use it, it would be simpler to drop this function (unless the review wants it, as you say).